### PR TITLE
Stop confusing the TF compiler with ModelOutput objects

### PR DESCRIPTION
**Pull request by Rocketknight1**
The `test_saved_model_creation` test was failing for BLIP with the rather unusual symptom that the `loss` key of one of the intermediate outputs had transformed into a strange generator object. I still don't know **why** this happened, but it requires the following:

1) TF compilation (doesn't happen in eager mode)
2) `ModelOutput` dicts with `loss` as the first, optional key
3) The `ModelOutput` dicts have to be returned internally and then used in a subsequent step, rather than returned as the last step of the outermost model

Since this is a nightmare zone, I'm going to work around the issue by just setting `return_dict` to `False` when calling the sub-model and get the tensors we need from the output tuple instead. This should be invisible for our users!

I also slipped a quick fix into the loss calculation to avoid potential NaNs from passing negative labels to one of the built-in TF loss functions. Even though all the negative-label positions should be masked, NaNs tend to persist (because `nan * 0 == nan`).

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for digging into and fixing this! 

Just some questions on my side to understand better the issue this is addressing. 

For the motivation itself, could you expand on: 
`The ModelOutput dicts have to be returned internally and then used in a subsequent step`
- What does it mean for the model output to be 'used'?

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 1]
```diff
@@ -1060,7 +1060,8 @@ def call(
             labels = labels[:, 1:]
             labels = tf.reshape(labels, (-1,))
             # Keras won't give us label smoothing for sparse CE, so we de-sparsify things here
-            one_hot_labels = tf.one_hot(labels, depth=self.config.vocab_size, dtype=tf.float32)
+            # Use relu to clamp masked labels at 0 to avoid NaN (we will be zeroing those out later anyway)
+            one_hot_labels = tf.one_hot(tf.nn.relu(labels), depth=self.config.vocab_size, dtype=tf.float32)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip_text.py:1064]: Yes, but the masked positions are set to 0 by the line `lm_loss *= masked_positions`! However, if we get a NaN value because we passed a negative label to an internal TF function, that won't be zeroed out correctly.

**Comment by Rocketknight1** [comment]
@amyeroberts sorry, let me explain what I meant with the model output being 'used'!

The issue occurs when the main `BlipModel` calls the `BlipText` sub-model, and the sub-model returns a `ModelOutput` which the main `BlipModel` reads values from. As far as I can tell, the issue does **not** arise if the `BlipText` model is called directly. So this seems to be some issue with the TF compiler not really understanding what kind of object a `ModelOutput` is when it has to compile through one inside a model `call()`.

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 2]
```diff
@@ -1171,20 +1171,27 @@ def call(
             attention_mask=attention_mask,
             encoder_hidden_states=image_embeds,
             labels=labels,
-            return_dict=return_dict,
+            return_dict=False,
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip.py:1174]: Regrettably yes - this bit is the bit that actually fixes the bug! The change to the loss logic turned out to be actually unrelated - I just spotted that there was a NaN risk there when I was trying to find the bug. Still, I think the change to it is worth keeping.

[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/blip/modeling_tf_blip.py:1174]: OK. As it's already done in BLIP and it's a pattern we also use for flax models I think it's fine to do :)

File: src/transformers/models/blip/modeling_tf_blip_text.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/blip/modeling_tf_blip_text.py:1064]: OK - works for me!

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing this tricky bug! 

Only request is for the slow test to be run before merging.

**Comment by Rocketknight1** [comment]
I ran the slow tests and all passed!
